File: python/aps/bundle.py
"""Agent Passport Bundle — aggregates passport, receipts, attestations, reputation, anchors."""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey

from . import crypto
from .passport import AgentPassport
from .receipt import WorkReceipt
from .attestation import Attestation
from .reputation import ReputationSummary


@dataclass
class AgentPassportBundle:
    passport: dict[str, Any] = field(default_factory=dict)
    work_receipts: list[dict[str, Any]] = field(default_factory=list)
    attestations: list[dict[str, Any]] = field(default_factory=list)
    reputation_summary: dict[str, Any] | None = None
    anchor_proofs: list[dict[str, Any]] = field(default_factory=list)
    proof: dict[str, Any] | None = None

    def to_dict(self, include_proof: bool = True) -> dict[str, Any]:
        d: dict[str, Any] = {
            "type": "AgentPassportBundle",
            "passport": self.passport,
            "work_receipts": self.work_receipts,
            "attestations": self.attestations,
            "reputation_summary": self.reputation_summary,
            "anchor_proofs": self.anchor_proofs,
        }
        if include_proof and self.proof is not None:
            d["proof"] = self.proof
        return d

    def sign(self, private_key: Ed25519PrivateKey) -> None:
        canonical = crypto.canonicalize_json(self.to_dict(include_proof=False))
        sig = crypto.ed25519_sign(private_key, canonical)
        now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
        agent_did = self.passport.get("id", "")
        self.proof = {
            "type": "Ed25519Signature2020",
            "created": now,
            "verificationMethod": agent_did + "#key-1",
            "proofPurpose": "assertionMethod",
            "proofValue": sig,
        }

    def verify(self, public_key: Ed25519PublicKey) -> bool:
        if self.proof is None:
            return False
        canonical = crypto.canonicalize_json(self.to_dict(include_proof=False))
        return crypto.ed25519_verify(public_key, canonical, self.proof["proofValue"])

    def verify_all(self, public_key: Ed25519PublicKey) -> bool:
        """Verify bundle signature, nested passport, and all receipts."""
        if not self.verify(public_key):
            return False

        # Verify passport
        p = AgentPassport.from_json(json.dumps(self.passport), validate=False)
        if p.proof and not p.verify(public_key):
            return False

        # Verify receipts
        for r_data in self.work_receipts:
            r = WorkReceipt.from_json(json.dumps(r_data), validate=False)
            if r.proof and not r.verify(public_key):
                return False

        return True

    def to_json(self) -> bytes:
        return crypto.canonicalize_json(self.to_dict())

    @classmethod
    def from_json(cls, data: bytes | str) -> "AgentPassportBundle":
        raw = json.loads(data) if isinstance(data, (bytes, str)) else data
        return cls(
            passport=raw.get("passport", {}),
            work_receipts=raw.get("work_receipts", []),
            attestations=raw.get("attestations", []),
            reputation_summary=raw.get("reputation_summary"),
            anchor_proofs=raw.get("anchor_proofs", []),
            proof=raw.get("proof"),
        )

File: typescript/src/bundle.ts
import { canonicalizeJson, ed25519Sign, ed25519Verify } from './crypto.js';
import { AgentPassport } from './passport.js';
import { WorkReceipt } from './receipt.js';
import type { Proof } from './types.js';

export interface AgentPassportBundleData {
  type: string;
  passport: Record<string, unknown>;
  work_receipts: Record<string, unknown>[];
  attestations: Record<string, unknown>[];
  reputation_summary: Record<string, unknown> | null;
  anchor_proofs: Record<string, unknown>[];
  proof?: Proof | null;
}

export class AgentPassportBundle {
  data: AgentPassportBundleData;

  constructor(data: Partial<AgentPassportBundleData> = {}) {
    this.data = {
      type: 'AgentPassportBundle',
      passport: data.passport || {},
      work_receipts: data.work_receipts || [],
      attestations: data.attestations || [],
      reputation_summary: data.reputation_summary || null,
      anchor_proofs: data.anchor_proofs || [],
      proof: data.proof || null,
    };
  }

  private toDict(includeProof = true): Record<string, unknown> {
    const d: Record<string, unknown> = {
      type: this.data.type,
      passport: this.data.passport,
      work_receipts: this.data.work_receipts,
      attestations: this.data.attestations,
      reputation_summary: this.data.reputation_summary,
      anchor_proofs: this.data.anchor_proofs,
    };
    if (includeProof && this.data.proof) {
      d.proof = this.data.proof;
    }
    return d;
  }

  async sign(privateKey: Uint8Array): Promise<void> {
    const canonical = canonicalizeJson(this.toDict(false));
    const sig = await ed25519Sign(privateKey, new TextEncoder().encode(canonical));
    const now = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
    const agentDid = (this.data.passport as Record<string, unknown>).id as string || '';
    this.data.proof = {
      type: 'Ed25519Signature2020',
      created: now,
      verificationMethod: agentDid + '#key-1',
      proofPurpose: 'assertionMethod',
      proofValue: sig,
    };
  }

  async verify(publicKey: Uint8Array): Promise<boolean> {
    if (!this.data.proof) return false;
    const canonical = canonicalizeJson(this.toDict(false));
    return ed25519Verify(publicKey, new TextEncoder().encode(canonical), this.data.proof.proofValue);
  }

  async verifyAll(publicKey: Uint8Array): Promise<boolean> {
    if (!(await this.verify(publicKey))) return false;

    // Verify passport
    const passport = AgentPassport.fromJson(JSON.stringify(this.data.passport), false);
    if (passport.data.proof) {
      if (!(await passport.verify(publicKey))) return false;
    }

    // Verify receipts
    for (const rData of this.data.work_receipts) {
      const receipt = WorkReceipt.fromJson(JSON.stringify(rData), false);
      if (receipt.data.proof) {
        if (!(await receipt.verify(publicKey))) return false;
      }
    }

    return true;
  }

  toJSON(): string {
    return canonicalizeJson(this.toDict());
  }

  static fromJSON(json: string): AgentPassportBundle {
    const raw = JSON.parse(json) as AgentPassportBundleData;
    return new AgentPassportBundle(raw);
  }
}
